### ISSUE TITLE
Decorators cause check-param-names to fail

### ISSUE THREAD

**Issue by Zamiell** [gajus/eslint-plugin-jsdoc#915]
## Expected behavior

```ts
/**
 * This is the description for my cool function.
 *
 * @param arg1 This is the number for my cool function.
 */
@MyDecorator
function foo(arg1: number) {}
```

This function should result in no errors.

## Actual behavior

```
@param "arg1" does not match an existing function parameter.
```

Removing the decorator fixes the problem.

## ESLint Config

```
"jsdoc/check-param-names": "warn"
```

**Comment by brettz9** [comment]
What version are you using? I am not getting an error for this at least with `master`.

**Comment by Zamiell** [comment]
I am using 39.3.6.

This is the specific function that I have, can you try to see if this one errors?

```ts
  /**
   * Helper function to register a custom pickup with the IsaacScript standard library. Use this
   * feature for custom pickups that are intended to be picked up by the player, like bombs and
   * keys.
   *
   * When IsaacScript detects that a player should be collecting the custom pickup, then the pickup
   * will be immediately removed, and an effect showing the pickup's respective `Collect` animation
   * will be spawned. (This emulates how a normal vanilla pickup would work.) After that, the
   * provided `collectFunc` will be fired. In this function, you will probably want to play a sound
   * corresponding to what kind of pickup it is (e.g. `SoundEffect.KEY_PICKUP_GAUNTLET`), as well as
   * do things like adding something to the player's inventory.
   *
   * Note that when you specify your custom pickup in the "entities2.xml" file, it should have a
   * type of "5" and be associated with an anm2 file that has a "Collect" animation.
   *
   * @param pickupVariantCustom The variant for the corresponding custom pickup.
   * @param subType The sub-type for the corresponding custom pickup.
   * @param collectFunc The function to run when the player collects this pickup.
   * @param collisionFunc Optional. The function to run when a player collides with the pickup.
   *                      Default is a function that always returns true, meaning that the player
   *                      will always immediately collect the pickup when they collide with it.
   *                      Specify this function if your pickup should only be able to be collected
   *                      under certain conditions.
   */
  @Exported
  public registerCustomPickup(
    pickupVariantCustom: PickupVariant,
    subType: int,
    collectFunc: (this: void, player: EntityPlayer) => void,
    collisionFunc: (this: void, player: EntityPlayer) => boolean = () => true,
  ): void {}
```

**Comment by brettz9** [comment]
For some reason, the arrow function default of `registerCustomPickup`'s `collisionFunc` is being iterated (`ArrowFunctionExpression`) by ESLint when there is a decorator but not without the decorator. A quick look at the AST doesn't make it appear to be `@typescript-eslint/parser`-related, so I wonder if it could be ESLint's traversal. Needs to be explored further, as I think we will need to know why they are inconsistent in the first place.

**Comment by brettz9** [comment]
This should now be fixed on `master` (via b001cd3deadb108952ffec5d8395f9e1cae82673 ). npm releases may still be having issues on our end as per https://github.com/gajus/eslint-plugin-jsdoc/issues/903#issuecomment-1267593452 . Closing as the issue is merged, but you can watch for the release.

**Comment by brettz9** [comment]
I might just mention that although I didn't determine precisely why the version with the decorator behaved differently, the issue in this case at least was that we shouldn't have been checking for JSDoc comments past an assignment pattern (the default expression needn't look further than the assignment to determine if a comment exists, e.g., the main function comment should not be considered as applicable to the default function value).

**Comment by Zamiell** [comment]
Awesome, thank you brett!
